src/jemalloc.c

```diff
@@ -2444,17 +2455,22 @@ je_malloc(size_t size) {
 	tcache_t *tcache = tsd_tcachep_get(tsd);
 	cache_bin_t *bin = tcache_small_bin_get(tcache, ind);
 	bool tcache_success;
-	void *ret = cache_bin_alloc_easy_reduced(bin, &tcache_success);
+	void *ret;
 
+	/*
+	 * We split up the code this way so that redundant low-water
+	 * computation doesn't happen on the (more common) case in which we
+	 * don't touch the low water mark.  The compiler won't do this
+	 * duplication on its own.
+	 */
+	ret = cache_bin_alloc_easy(bin, &tcache_success);
 	if (tcache_success) {
-		thread_allocated_set(tsd, allocated_after);
-		if (config_stats) {
-			bin->tstats.nrequests++;
-		}
-
-		LOG("core.malloc.exit", "result: %p", ret);
-
-		/* Fastpath success */
+		fastpath_success_finish(tsd, allocated_after, bin, ret);
+		return ret;
+	}
+	ret = cache_bin_alloc(bin, &tcache_success);
```

Comment:
So basically this extra `cache_bin_alloc()` check is how you got the performance win (so that this "second-order" common case won't need to be routed to the slow path)? And an essential part of why it can work is that the extra check can now be accomplished without costing extra registers, which is credited to the restructuring? Otherwise, logic wise, the current computation is pretty much equivalent to what was the case before https://github.com/jemalloc/jemalloc/commit/05681e387a3202567ff95528dbc460e92e031a3c?

Reply:
Right -- I think that the actual pointer reserved and the layout of pointers in the cache bin array should be almost completely unchanged with this refactor.